### PR TITLE
TP sharding v2

### PR DESCRIPTION
**Pull request by Narsil**
Second update version of TP sharding.

- This WAS NOT tested for accuracy (beacuse I didn't implement all_reduce)
- Currently waiting for a simple API to get the cuda storage so I can call `all_reduce` directly and recreate a tensor from that. (That way I don't just `pub` everything)

I tried to keep the modifications minimal in `var_builder` for now as this is not the purpose of this PR.

### CODE REVIEW

File: candle-examples/examples/llama_multiprocess/model.rs [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+        self.comm.all_reduce(s, &mut dst, &ReduceOp::Sum).unwrap();
+        let dst = candle::CudaStorage::wrap_cuda_slice(dst, dev);
+        Ok((dst, l.shape().clone()))
+    }
+}
+
+fn all_reduce_sum(x: &Tensor, comm: &Rc<Comm>) -> Result<Tensor> {
+    x.custom_op1(AllReduce { comm: comm.clone() })
+}
+
```
Comment (Narsil) [inline candle-examples/examples/llama_multiprocess/model.rs:74]: This is the core of the new thing.

+ TP Row/Col

File: candle-nn/src/var_builder.rs [hunk 2]
```diff
@@ -137,6 +135,58 @@ impl<'a> VarBuilder<'a> {
...
+                let view = safetensors[*index].tensor(&path)?;
+                let dtype = view.dtype();
+                let mut shape = view.shape().to_vec();
+                let size = shape[dim];
+                let block_size = size / world_size;
+                let start = rank * block_size;
+                let stop = (rank + 1) * block_size;
+
+                let iterator = if dim == 0 {
+                    view.slice(start..stop).unwrap()
```
Comment (LaurentMazare) [inline candle-nn/src/var_builder.rs:172]: Why unwrap rather than propagating the error?

File: candle-nn/src/var_builder.rs [hunk 3]
```diff
@@ -137,6 +135,58 @@ impl<'a> VarBuilder<'a> {
...
+                } else if dim == 1 {
+                    view.slice((.., start..stop)).unwrap()
+                } else {
+                    unimplemented!("Get sharded on dimensions != 0 or 1");
```
Comment (LaurentMazare) [inline candle-nn/src/var_builder.rs:176]: You can use `Error::bail!` to have a proper error if we don't plan on supporting this anytime soon.

File: candle-nn/src/var_builder.rs [hunk 4]
```diff
@@ -137,6 +135,58 @@ impl<'a> VarBuilder<'a> {
 }
 
 impl<'a> VarBuilder<'a> {
+    pub fn get_sharded(
+        &self,
+        tensor_name: &str,
+        dim: usize,
+        rank: usize,
+        world_size: usize,
```
Comment (LaurentMazare) [inline candle-nn/src/var_builder.rs:143]: It's a bit unclear what `world_size` means here, add a comment or some documention to the function?

File: candle-nn/src/var_builder.rs [hunk 5]
```diff
@@ -137,6 +135,58 @@ impl<'a> VarBuilder<'a> {
...
+                    }
+                    .bt()
+                })?;
+
+                let view = safetensors[*index].tensor(&path)?;
+                let dtype = view.dtype();
+                let mut shape = view.shape().to_vec();
+                let size = shape[dim];
+                let block_size = size / world_size;
+                let start = rank * block_size;
```
Comment (LaurentMazare) [inline candle-nn/src/var_builder.rs:168]: It's a bit tricky to know what is a size in number of elements or a size in bytes here, maybe rename to `start_in_bytes` or something like this? (`size` seems to be in elements)

File: candle-nn/src/var_builder.rs (hunk 2 above)
Comment (Narsil) [reply candle-nn/src/var_builder.rs:172]: Fixed.

File: candle-nn/src/var_builder.rs (hunk 3 above)
Comment (Narsil) [reply candle-nn/src/var_builder.rs:176]: Fixed

File: candle-nn/src/var_builder.rs (hunk 5 above)
Comment (Narsil) [reply candle-nn/src/var_builder.rs:168]: Everything is in number of elements. The size in bytes is taken care of in `safetensors`.

File: candle-nn/src/var_builder.rs (hunk 4 above)
Comment (Narsil) [reply candle-nn/src/var_builder.rs:154]: I updated with a rustdoc on top.

`rank` and `world_size` are definitely names borrowed from torch distributed.
We could rename them to something like `block_id` , `n_blocks` or something along those lines.

File: candle-nn/src/var_builder.rs (hunk 5 above)
Comment (Narsil) [reply candle-nn/src/var_builder.rs:187]: Done that.

**Review by LaurentMazare — APPROVED** [review]

**Pull request merged**